File: solutions/199_binary_tree_right_side_view.cpp
#include "aux.cpp"
/**
Given a binary tree, imagine yourself standing on the right side of it, return
the values of the nodes you can see ordered from top to bottom. Example:
Input:[1,2,3,null,5,null,4]
Output:[1, 3, 4]
Explanation:
   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---

 https://leetcode.com/problems/binary-tree-right-side-view/
 **/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
class Solution {
public:
  vector<int> rightSideView(TreeNode *root) {
    vector<int> res;
    tranversal(root, 1, res);
    return res;
  }

  void tranversal(TreeNode *root, int depth, vector<int> &res) {
    if (nullptr == root)
      return;
    if (res.size() < depth)
      res.emplace_back(root->val);
    res[depth - 1] = root->val;
    tranversal(root->left, depth + 1, res);
    tranversal(root->right, depth + 1, res);
  }
};

int main() {
  Solution s;
  TreeNode *root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(3);
  root->right->right = new TreeNode(4);
  root->left->right = new TreeNode(5);

  say(s.rightSideView(root));
}

File: _site/solutions/199_binary_tree_right_side_view.rb
# Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.
# Example:
# Input:[1,2,3,null,5,null,4]
# Output:[1, 3, 4]
# Explanation:
#    1            <---
#  /   \
# 2     3         <---
#  \     \
#   5     4       <---
#
#  https://leetcode.com/problems/binary-tree-right-side-view/
require './aux.rb'

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
  res = []
  trav = lambda do |r, depth|
    return if r.nil?
    res[depth] = r.val
    trav.call(r.left, depth + 1)
    trav.call(r.right, depth + 1)
  end
  trav.call(root, 0)
  res
end

arr = [1, 2, 3, nil, 5, nil, 4]
root = construct_tree(arr)
p right_side_view(root)
